File: tests/mobi/hsz/idea/gitignore/completion/CompletionTest.java
package mobi.hsz.idea.gitignore.completion;

import com.intellij.openapi.application.Result;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase;
import mobi.hsz.idea.gitignore.file.type.kind.GitFileType;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class CompletionTest extends LightPlatformCodeInsightFixtureTestCase {
    public void testSimple() {
        myFixture.getTempDirFixture().createFile("fileName.txt");
        doTest("fileN<caret>", "fileName.txt<caret>");
    }

    public void testCurrentDirectoryAlias() throws IOException {
        new WriteCommandAction(getProject()) {
            @Override
            protected void run(@NotNull Result result) {
                myFixture.getTempDirFixture().createFile("fileName.txt");
            }
        }.execute();

        doTest("./fileN<caret>", "./fileName.txt<caret>");
    }

    public void testInHiddenDirectory() throws IOException {
        new WriteCommandAction(getProject()) {
            @Override
            protected void run(@NotNull Result result) throws IOException {
                myFixture.getTempDirFixture().findOrCreateDir(".hidden").createChildData(this, "fileName.txt");
            }
        }.execute();

        doTest(".hidden/fileN<caret>", ".hidden/fileName.txt<caret>");
    }

    public void testInGlobDirectory() throws IOException {
        new WriteCommandAction(getProject()) {
            @Override
            protected void run(@NotNull Result result) throws IOException {
                myFixture.getTempDirFixture().findOrCreateDir("glob1").createChildData(this, "fileName1.txt");
                myFixture.getTempDirFixture().findOrCreateDir("glob2").createChildData(this, "fileName2.txt");
            }
        }.execute();

        doTestVariants("*/fileN<caret>", "fileName1.txt", "fileName2.txt");
    }

    public void testNegation() throws IOException {
        myFixture.getTempDirFixture().createFile("fileName.txt");
        doTest("!fileNa<caret>", "!fileName.txt<caret>");
    }

    private void doTest(@NotNull String beforeText, @NotNull String afterText) {
        myFixture.configureByText(GitFileType.INSTANCE, beforeText);
        myFixture.completeBasic();
        myFixture.checkResult(afterText);
    }

    private void doTestVariants(@NotNull String beforeText, String... variants) {
        myFixture.configureByText(GitFileType.INSTANCE, beforeText);
        myFixture.completeBasic();
        assertContainsElements(myFixture.getLookupElementStrings(), variants);
    }
}

File: src/test/kotlin/mobi/hsz/idea/gitignore/completion/CompletionTest.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package mobi.hsz.idea.gitignore.completion

import com.intellij.openapi.command.WriteCommandAction.writeCommandAction
import com.intellij.testFramework.UsefulTestCase
import mobi.hsz.idea.gitignore.Common
import mobi.hsz.idea.gitignore.file.type.IgnoreFileType
import mobi.hsz.idea.gitignore.file.type.kind.GitFileType
import java.io.IOException

class CompletionTest : Common<Any?>() {

    fun testSimple() {
        myFixture.tempDirFixture.createFile("fileName.txt")
        doTest("fileN<caret>", "fileName.txt<caret>")
    }

    @Throws(IOException::class)
    fun testCurrentDirectoryAlias() {
        writeCommandAction(project).run<Throwable> {
            myFixture.tempDirFixture.createFile("fileName.txt")
        }
        doTest("./fileN<caret>", "./fileName.txt<caret>")
    }

    @Throws(IOException::class)
    fun testInHiddenDirectory() {
        writeCommandAction(project).run<Throwable> {
            myFixture.tempDirFixture.findOrCreateDir(".hidden").createChildData(this, "fileName.txt")
        }
        doTest(".hidden/fileN<caret>", ".hidden/fileName.txt<caret>")
    }

    @Throws(IOException::class)
    fun testInGlobDirectory() {
        writeCommandAction(project).run<Throwable> {
            myFixture.tempDirFixture.findOrCreateDir("glob1").createChildData(this, "fileName1.txt")
            myFixture.tempDirFixture.findOrCreateDir("glob2").createChildData(this, "fileName2.txt")
        }
        doTestVariants("*/fileN<caret>", "fileName1.txt", "fileName2.txt")
    }

    @Throws(IOException::class)
    fun testNegation() {
        myFixture.tempDirFixture.createFile("fileName.txt")
        doTest("!fileNa<caret>", "!fileName.txt<caret>")
    }

    private fun doTest(beforeText: String, afterText: String) {
        myFixture.configureByText(IgnoreFileType.INSTANCE, beforeText)
        myFixture.completeBasic()
        myFixture.checkResult(afterText)
    }

    private fun doTestVariants(beforeText: String, vararg variants: String) {
        myFixture.configureByText(IgnoreFileType.INSTANCE, beforeText)
        myFixture.completeBasic()
        UsefulTestCase.assertContainsElements(myFixture.lookupElementStrings!!.toList(), *variants)
    }
}
